src/goose/synopsis/toolkit.py

```diff
@@ -157,6 +136,207 @@ def patch_file(self, path: str, before: str, after: str) -> str:
         self.notifier.log("")
         return "Succesfully replaced before with after."
 
+    def _save_file_history(self, patho: Path) -> None:
+        """Save the current content of the file to history for undo functionality."""
+        content = patho.read_text() if patho.exists() else ""
+        self._file_history[str(patho)] = content
+
+    def _undo_edit(self, path: str) -> str:
+        """Undo the last edit made to a file."""
+        patho = system.to_patho(path)
+
+        if not patho.exists():
+            raise ValueError(f"The file {path} does not exist.")
+
+        if str(patho) not in self._file_history:
+            raise ValueError(f"No edit history available to undo changes on {path}.")
+
+        # Restore the previous content from history
+        previous_content = self._file_history.pop(str(patho))
+        patho.write_text(previous_content)
+
+        # Remember the file in system
+        system.remember_file(path)
+
+        get_language(path)
+        md_content = f"Undo edit for {path}."
+
+        self.notifier.log("")
+        self.notifier.log(Rule(RULEPREFIX + path, style=RULESTYLE, align="left"))
+        self.notifier.log(Markdown(md_content))
+        self.notifier.log("")
+
+        return f"Successfully undid the last edit on {path}"
+
+    @tool
+    def text_editor(
```

Comment:
there's a lot of conflated concerns in a single method. it's a little tough to read and would even harder to test. how about trying out a dispatch pattern?

```python3
command_dispatch: dict[str, callable] = {
    "create": self._create_file,
    "insert": self._insert_string,
    "undo_edit": self._undo_edit,
     ...
}
```

Reply:
we should possibly consider creating a text editor abstraction given the complex interactions we have here. maybe these things would play well with the editor integrations people have made plugins for?

Reply:
like the suggestion for dispatcher pattern. i opened another PR cause i also batched up the changes for bash and process manager: https://github.com/block/goose/pull/191

Reply:
@lamchau I think this is the approach that claude took and it seems to work well - a smaller number of tools I think is cognitively more pleasant vs LLM needing to decide which tool to use (also lets us put more smarts in the tool vs relying on LLM always to choose the right way) - so I think this general approach feels right?

Reply:
@michaelneale not sure i follow? i think the logic itself is sound just tidying it up